byterun/memory.c

```diff
@@ -97,6 +97,49 @@ CAMLexport int caml_atomic_cas_field (value obj, int field, value oldval, value
   }
 }
 
+
+/* FIXME: is __sync_synchronize a C11 SC fence? Is that enough? */
+
+CAMLprim value caml_atomic_load (value ref)
+{
+  if (Is_young(ref)) {
+    return Op_val(ref)[0];
+  } else {
+    CAMLparam1(ref);
+    CAMLlocal1(v);
+    __sync_synchronize();
+    caml_read_field(ref, 0, &v);
+    __sync_synchronize();
+    CAMLreturn (v);
+  }
+}
+
+CAMLprim value caml_atomic_store (value ref, value v)
+{
+  __sync_synchronize();
+  caml_modify_field(ref, 0, v);
+  __sync_synchronize();
+  return Val_unit;
+}
+
+CAMLprim value caml_atomic_cas (value ref, value oldv, value newv)
+{
+  value* p = Op_val(ref);
+  if (Is_young(ref)) {
+    if (*p == oldv) {
+      *p = newv;
+      write_barrier(ref, 0, newv);
```

Comment:
Is it safe to call `write_barrier` after the value has been updated? The current write barrier assumes that `write_barrier` is invoked before the update: https://github.com/ocamllabs/ocaml-multicore/blob/master/byterun/memory.c#L33. 
Since the write barrier only expects to read the old value in the case of young ref, we should be able to safely swap this line with the previous line. 

I believe `caml_atomic_cas_field` might also be incorrect due to this reason and needs a similar change.